### PR TITLE
FIX: avoid race condition when setting user status

### PR DESCRIPTION
**Pull request by AndrewPrigorshnev**
We caught it in logs, race condition led to this error:

    ActiveRecord::RecordNotUnique 
    (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "user_statuses_pkey"
    DETAIL:  Key (user_id)=(15) already exists.)

The reason the problem happened was that we were checking if a user has status and if not inserting status:

    if user_status
      ...
    else
      self.user_status = UserStatus.create!(status)
    end

The problem is that it's possible that another request will insert status just after we check if status exists and just before our request call `UserStatus.create!(status)`. Using `upsert` fixes the problem because under the hood `upsert` generates the only SQL request that uses "INSERT ... ON CONFLICT DO UPDATE". So we do everything in one SQL query, and that query takes care of resolving possible conflicts.

### CODE REVIEW

**Comment by danielwaterworth** [comment]
@AndrewPrigorshnev, I'm not convinced this fixes the issue, could you explain why this works?

**Comment by AndrewPrigorshnev** [comment]
@danielwaterworth, Oh, this won't work, actually. I had this idea yesterday and opened a PR without a lot of thinking, pretty silly :)

I've reimplemented it using `upsert`, it generates `INSERT ... ON CONFLICT DO UPDATE`, that should resolve the issue:

image

One drawback of using `upsert` is that it doesn't perform validations, so we have to do it ourselves.

Testing this is pretty tricky I think we can do it without tests, I think `upsert` is a good bet, what do you think?

File: app/models/user.rb [hunk 1]
```diff
@@ -2174,6 +2177,11 @@ def self.ensure_consistency!
       )
     SQL
   end
+
+  def validate_status!(status)
+    model = UserStatus.new(status)
+    raise(ActiveRecord::RecordInvalid.new(model)) if model.invalid?
+  end
```
Comment (jjaffeux) [inline app/models/user.rb:2184]:
```
model = UserStatus.new(status)
model.validate!
model
```

**Review by jjaffeux — APPROVED** [review]
:+1:

**Comment by jjaffeux** [comment]
Can you try to explain a little bit in the desc what was causing the exception please?

**Pull request merged**